libtool: fix /usr/bin/file impurity

a-m-joseph opened the pull request:
## Problem

In libtool<=2.4.6 the libtool.m4 script assumes that `file` is available, and can be found at `/usr/bin/file` ([this path is hardwired](https://git.savannah.gnu.org/cgit/libtool.git/tree/m4/libtool.m4?h=v2.4.6&id=f09d00cbcf924c378573163e244fffeb8d28005f#n1346)).  This fails for sandboxed builds.  This script is vendored into the `./configure` scripts of an enormous number of packages.

The use of /usr/bin/file in ./configure scripts affects sandboxed cross-builds to a `x86_64-linux` hosts as well as all sandboxed builds     (cross and native) to  `x86_64-freebsd`, `*-hpux`, `*-irix`,     `mips64*-linux`, `powerpc*-linux`, `s390x-linux`, `s390x-tpf`,  `sparc-linux`,     and `*-solaris`.

## (Immediate) Motivation

This is [currently preventing](https://logs.nix.ci/?attempt_id=eeeefd2c-7df3-4852-97e9-c2063960f67b&key=nixos%2Fnixpkgs.166484) the generation of Hydra-built `bootstrap-files` for mips64el-linux-gnuabi64, and will affect many of the other new platforms we try to bootstrap on. 

The build of the `pcre` package (and a few others) emits these lines when its configure script is run:

```
  ./configure: line 9843: /usr/bin/file: No such file or directory
  ./configure: line 9851: /usr/bin/file: No such file or directory
  ./configure: line 9859: /usr/bin/file: No such file or directory
  ...
  checking whether to build shared libraries... no
```

The build does not fail until much later, when make-bootstrap-files.nix tries to make a copy libpcre.so, which was never built.

## Solution

An earlier revision of this PR tried to use fancy `stdenv` adapters to fix this problem without causing mass rebuilds.  SuperSandro2000's [suggestion](https://github.com/NixOS/nixpkgs/pull/166879#discussion_r847783773) to simply apply the fix everywhere seems like it will be more maintainable.

Currently this PR adds a new top-level expression `replaceUsrBinFile` which will check to see if the package being built has a vendored `libtool.m4` older than 2.4.7.  If it does, then it will `substituteInPlace ./configure --replace /usr/bin/file ${buildPackages.file}/bin/file`.  If it does *not* find this, it will abort the build with an error message.  This ensures that maintainers will remove the `replaceUsrBinFile` as soon as upstream regenerates their autoconf with an unaffected libtool >= 2.4.7.

This commit also adds `replaceUsrBinFile` to `nativeBuildInputs` of a large number (but not all) of the packages that need it.


###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [X] mips64el-linux cross from x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

NickCao commented:
Can't this be a part of autoreconfHook?

a-m-joseph commented:
> Can't this be a part of autoreconfHook?

Unfortunately autoreconfHook isn't called by default.  So a stdenv adapter would still be needed to add an extra dependency.

Calling autoreconfHook instead of fixPathToUsrBinFileInConfigureScript could be done in theory, but in practice autoreconfHook breaks a lot of packages -- it uses the latest autotools instead of whatever version was used when the package was created.  The ./configure script is basically a vendored dependency.

The breakage caused by blindly autoreconfHook'ing things is why I prefer this PR over  #166722.  I tried getting `pkgsCross.mips64el-linux-gnuabi64.nix_2_4` to build with that one and it just turned into a huge mess.  With this PR I just (minutes ago) finished cross-compiling `nix_2_4` and everything else that runs on my routers, quite a lot of software, and the only additional patch needed is for `libidn2`.

Mindavi commented:
So this is a temporary solution until libtool is upgraded? At least it's nice to have this solved in the short term.

a-m-joseph commented:
> So this is a temporary solution until libtool is upgraded?

No.

Read the section titled "implementation".  There are two changes, only one of them can be removed when libtool is upgraded.  The other one needs to stay due to vendoring in `./configure` scripts.

a-m-joseph commented:
Pushed three additional patches.  One just expands a comment with more detail, another factors out `needsLibtoolUsrBinFileFix` so it can be referenced by package expressions, and the third uses it to prevent the `curl` expression from breaking the `./configure` script on platforms that need `file` to detect library flags.

As a side effect of the third patch, `file` is now part of the "stuff needed to bootstrap fetchurl" clique.  I've added the boilerplate comment to declare this fact.

Perhaps we should consider including `file` in bootstrap-tools later on.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/nixlang-how-do-you-find-all-uses-of-a-declaration/18369/14

Ericson2314 commented:
I was talking to @alyssais about this. She pointed out https://discourse.nixos.org/t/rfc-libtool-updates-and-improved-nix-nixos-support/15660 this to me too, which is very exciting.

I would like to move closer to `autoreconfPhase` by default, for all platforms.

a-m-joseph commented:
> I was talking to @alyssais about this. She pointed out https://discourse.nixos.org/t/rfc-libtool-updates-and-improved-nix-nixos-support/15660 this to me too, which is very exciting.

Yes, that's certainly good news!  But just like the bugfix in libtool-2.4.7, it won't affect the hundreds (or thousands?) of packages that have an old `libtool.m4` vendored into their `./configure`.  This brings us to your other point:

> I would like to move closer to `autoreconfPhase` by default, for all platforms.

I would like that as well, but the amount of breakage it will cause is mind-boggling.  So many packages fail to build when you add `autoreconfHook`, it's depressing -- unless there is some secret magic trick I'm unaware of.

I hope this PR does not end up blocked on some hoped-for future `autoreconfPhase`-by-default.  This PR is the one and only thing ([proof](https://github.com/NixOS/nixpkgs/pull/166722#issuecomment-1085244318) from ofborg) preventing Hydra from building a mips64el bootstrap-files whose hash can be committed to `pkgs/stdenv/linux/bootstrap-files/`.  Making that wait for all the `autoreconfHook` bugs in all of nixpkgs to get fixed would be pretty soul-crushing.

a-m-joseph commented:
(I'm currently tweaking this to reduce the amount of rebuilds it causes)

Commit:
libtool: fix /usr/bin/file impurity

libtool's libtool.m4 script assumes that `file` is available, and can
be found at `/usr/bin/file` (this path is hardwired).  Furthermore,
the script with this assumption is vendored into the ./configure
scripts of an enormous number of packages.

The use of /usr/bin/file in ./configure scripts only affects sandboxed
cross-builds to a x86_64-linux host as well as any sandboxed build
(cross or native) to these hosts: x86_64-freebsd, *-hpux, *-irix,
mips64*-linux, powerpc*-linux, s390x-linux, s390x-tpf, sparc-linux,
and *-solaris.

This commit adds `replaceUsrBinFile`, which will check the ./configure
script to see if it vendors a pre-version-2.4.7 copy of libtool.m4.
If it does, it will `substituteInPlace` the use of `/usr/bin/file`
with `${buildPackages.file}/bin/file`.

If a pre-version-2.4.7 copy of libtool.m4 is *not* vendored then
replaceUsrBinFile will cause the build to fail -- this ensures that
replaceUsrBinFile gets removed as upstream packages gradually upgrade
to a fixed version of libtool and re-run autoconf.

This commit allows the following commands to complete, which should
enable Hydra to produce bootstrap-files for mips64el:

```
  nix-build \
    --option sandbox true \
    --option sandbox-fallback false \
    pkgs/top-level/release-cross.nix \
    -A bootstrapTools.mips64el-linux-gnuabi64.build

  nix-build \
    --option sandbox true \
    --option sandbox-fallback false \
    . \
    -A pkgsCross.mips64el-linux-gnuabi64.nix_2_4
```

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

Commit:
libtool: 2.4.6 -> 2.4.7

libtool2-macos11.patch has been merged upstream as 9e8c88251

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

a-m-joseph commented:
> 2\. fix `libtool.m4` in the libtool package (this [is fixed upstream](https://git.savannah.gnu.org/cgit/libtool.git/commit/?h=branch-1-5&id=da2e352735722917bf0786284411262195a6a3f6)  in libtool 2.4.7)

#167071 is a PR to upgrade to libtool 2.4.7.  Unfortunately it causes downstream build failures with `--option sandbox true`.  I have a mass-rebuild running with `--option sandbox false` and it's going very well, so the sandboxing issue is probably the only obstacle.

If #167071 merges first then the changes to `pkgs/development/tools/misc/libtool/libtool2.nix` in this PR can be dropped.  However the rest of this PR remains necessary regardless of the libtool upgrade.

a-m-joseph commented:
> So many packages fail to build when you add `autoreconfHook`, it's depressing -- unless there is some secret magic trick I'm unaware of.

Here's a great example: https://github.com/NixOS/nixpkgs/pull/163401#issuecomment-1072040524

SuperSandro2000 commented on pkgs/development/tools/misc/libtool/libtool2.nix:
```diff
@@ -36,6 +37,29 @@ stdenv.mkDerivation rec {
+  #
+  #   https://git.savannah.gnu.org/cgit/libtool.git/commit/?h=branch-1-5&id=da2e352735722917bf0786284411262195a6a3f6
+  #
+  preConfigure = if stdenv.buildPlatform == stdenv.targetPlatform
+                 then null
+                 else ''
+                 substituteInPlace m4/libtool.m4     --replace "/usr/bin/file" "${buildPackages.file}/bin/file"
+                 substituteInPlace configure         --replace "/usr/bin/file" "${buildPackages.file}/bin/file"
+                 substituteInPlace libltdl/configure --replace "/usr/bin/file" "${buildPackages.file}/bin/file"
+                 '';
```
```suggestion
  preConfigure = lib.optionalString (stdenv.buildPlatform != stdenv.targetPlatform) ''
    for file in configure libltdl/configure m4/libtool.m4; do
      substituteInPlace $file  --replace "/usr/bin/file" "${buildPackages.file}/bin/file"
    done
  '';
```

that should be equivalent but shorter and sleaker.

SuperSandro2000 commented on pkgs/development/libraries/libidn2/default.nix:
```diff
@@ -20,9 +22,15 @@ stdenv.mkDerivation rec {
   # documentation tools are required.
-  nativeBuildInputs = optionals stdenv.isDarwin [ help2man texinfo ];
+  nativeBuildInputs = optionals stdenv.isDarwin [ help2man texinfo ]
+  # libidn2's Makefile "helpfully" forces an autoreconf if it notices changes to the ./configure script
+  ++ lib.optionals (stdenv.hostPlatform != stdenv.buildPlatform) [ autoreconfHook buildPackages.pkgconfig];
```
That is pretty standard across many projects. Also pkgconfig should be pkg-config.

SuperSandro2000 commented on lib/systems/default.nix:
```diff
@@ -68,6 +68,28 @@ rec {
       useAndroidPrebuilt = false;
       useiOSPrebuilt = false;
 
+      needsLibtoolUsrBinFileFix = isCross: with final;
```
Suggestion: How about we always fix this to not needing to keep track of this and in projects? Because I don't think that anytime soon this is fixed *everywhere*.

a-m-joseph replied:
@SuperSandro2000, thank you so much for taking the time to review this PR (and my others too, but this one especially).  I really appreciate it; this is the PR I was afraid would get stuck in limbo forever.

> Suggestion: How about we always fix this to not needing to keep track of this and in projects?

I would prefer this too.  I was trying to minimize rebuilds, but in hindsight maintainability matters more.

If the fix is applied on all platforms then this definitely needs to go into staging.  The good news: if this goes into staging it can build off of #167071, which will make this PR much shorter.

I will incorporate your other feedback, switch branches, and remove the parts that are no longer necessary after #167071.

Again, thank you so much for wading into this complicated issue and taking the time to understand what's going on here!

SuperSandro2000 reviewed (commented):
Not a big fan that we add another package to bootstrapping.

The names you have chosen are horrible long. I don't think anyone can remember them. Maybe we can passthru them for autoconf and autoreconfHook so that you can easily access them.

a-m-joseph commented on pkgs/top-level/all-packages.nix:
```diff
@@ -180,6 +180,22 @@ with pkgs;
 
+  # libtool.m4 prior to version 2.4.7 invokes "/usr/bin/file", and is vendored into many ./configure scripts
+  fixPathToUsrBinFileInConfigureScriptIfNecessary =
+    if stdenv.hostPlatform.needsLibtoolUsrBinFileFix (stdenv.hostPlatform != stdenv.buildPlatform)
+    then [ fixPathToUsrBinFileInConfigureScript ]
+    else [ ];
```
Hrm, github is not letting me commit your change ("This diff has recently been updated.") but I will integrate it manually.

SuperSandro2000 replied:
> This diff has recently been updated.

I sometimes get that, too. I don't know why but it is always fine to just copy the suggestions and apply it locally and then push it.

a-m-joseph commented:
Okay, I think this should work.  Running some rebuilds overnight to be sure.

The new approach applies the `substituteInPlace` regardless of platform.

This let me simplify the top-level addition, which is now called just `replaceUsrBinFile`.  It now checks to see if it is necessary, and fails the build if it is applied unnecessarily.  This will ensure that this new extra dependency gets removed when we upgrade packages to versions that no longer need the hack.

SuperSandro2000 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -181,6 +181,26 @@ with pkgs;
     '');
 
+  # libtool.m4 prior to version 2.4.7 invokes "/usr/bin/file", and is vendored into many ./configure scripts
+  replaceUsrBinFile =
```
Maybe we could convert this to a setupHook in which case we would never need to call it manually? We would need to remove the exit part but it would make it a lot less work to maintain.

https://github.com/NixOS/nixpkgs/pull/166486/files#diff-7510b47653eab58ead900c36ffc4881682b881b87c0e7b7016d1f3947dee495cR50

sternenseemann replied:
I think it would be logical to make this part of `fixLibtool` in `pkgs/stdenv/generic/setup.sh`.

a-m-joseph replied:
I would love to do that but I can't figure out how to thread the reference to ${buildPackages.file} down into `setup.sh`.  Also it would create a circular dependency in many cases, wouldn't it?

By the way, speaking of crazy circular dependencies, the `./configure` script that ships with `file-5.41.tar.gz` tries to call `/usr/bin/file`!  The package *ships with* a circular dependency!  Talk about crazy...

a-m-joseph replied:
> Maybe we could convert this to a setupHook

Thanks I will try that.

a-m-joseph replied:
> the `./configure` script that ships with `file-5.41.tar.gz` tries to call `/usr/bin/file`! The package _ships with_ a circular dependency!

> I would love to do that but I can't figure out how to thread the reference to `${buildPackages.file}` down into `setup.sh`. Also even if I could do that, it would create a circular dependency in many cases, wouldn't it?

I should add that these are two arguments for considering putting `file` in the bootstrap-tools.  I know that adding more stuff to the bootstrap-tools is never an easy decision, but when you have packages that depend on themselves (like a C compiler) and which 99% of the ecosystem has a vendored dependency upon, well... that's sort of what the bootstrap-files is for.

Then we could have `fixLibtool()` replace `/usr/bin/file` with `file`, since the bootstrap-tools `bin` directory is always in the `$PATH`.  So there wouldn't be any need to thread the package reference down into there through the nixlang code.

It's ugly, but so is the problem that it fixes.

a-m-joseph replied:
See #168413 for a much, much simpler solution that simply (a) patches configure scripts unconditionally and (b) includes `file` in stdenv's `common-path.nix`.

I'm running a rebuild-the-world on that PR, it'll take all day.

a-m-joseph commented:
Closed in favor of #168413, which is much much simpler and more maintainable -- at the cost of having to rebuild everything.
